Core/Src/main.c

```diff
@@ -143,7 +154,30 @@ void SystemClock_Config(void)
 }
 
 /* USER CODE BEGIN 4 */
+void TIM1_CC_IRQHandler(void)
+{
+  // Interrupt handler code here
+  // reached desired pwm pulse width, so set PWM pin low
+  HAL_GPIO_WritePin(GPIOA, GPIO_PIN_8, GPIO_PIN_RESET);
+}
 
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+  if (htim->Instance == TIM1)
+  {
+	HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8); // pull down Chip Select
+	uint8_t tx_data[3] = {0x01, 0x80, 0x00};
+	uint8_t rx_data[3] = {0};
+
+	HAL_SPI_TransmitReceive(&hspi1, tx_data, rx_data, 3, 1000);
+	ADC_value = ((rx_data[1] & 0x03) << 8) | rx_data[2]; // read ADC value
+
+	pwm_pulse_width = (ADC_value * (__HAL_TIM_GET_AUTORELOAD(&htim1) + 1)) / 4096;
+	TIM1->CCR1 = pwm_pulse_width;
```

Comment:
you should be able to use a different function instead of setting the CCR directly :)